p4-16/psa/PSA.mdk

```diff
@@ -523,34 +523,71 @@ extern Checksum<W> {
 
 ## Counters
 
-Counters are a simple mechanism for keeping statistics about the
-packets that trigger a table in a Match Action unit.
+Counters are a mechanism for keeping statistics.  The control plane
+can read counter values.  A P4 program cannot read counter values,
+only update them.  If you wish to implement a feature involving
+sequence numbers, for example, use Registers instead (Section
+[#sec-registers]).
+
+Direct counters are counters associated with a particular P4 table,
+and are implemented by the extern `DirectCounter`.  There are also
+'indirect' counters, which are implemented by the extern `Counter`.
+The primary differences between direct counters and indirect counters
+are:
+
+- Number of independently updatable counter values:
+  - A single instantiation of a direct counter always contains as many
+    independent counter values as the number of entries in the table
+    with which it is associated.
+  - You must specify the number of independent counter values for an
+    indirect counter when instantiating it.  This number of counters
+    need not be the same as the size of any P4 table.
+- Where counter updates are allowed in the P4 program:
+  - For a direct counter, you may only invoke its `count` method from
+    inside the actions of the table with which it is associated, and
+    this always updates the counter value associated with the matching
+    table entry.
+  - For an indirect counter, you may invoke its `count` method
+    anywhere in the P4 program where extern object method invocations
+    are permitted (e.g. inside actions, or directly inside a control's
+    `apply` block), and every such invocation must specify the index
+    of the counter value to be updated.
```

Comment:
Thanks for the thoughtful comments here.  My apologies if I'm a little too argumentative on some of these points at times.

I do like the goal of a PSA that defines externs, rules for using those externs in P4 programs such that those P4 programs should be portable across PSA implementations, as well as control plane APIs that are portable across PSA implementations.  I'd be happy if that were some kind of 'mission statement' for the PSA.

While I consider it slightly unfortunate if we have small differences between PSA implementations, such as using different values for packet lengths on counter and meter updates, I also think most people could be content with such differences, more often than not.